### PR TITLE
[codex] Add policy lifecycle aggregate

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- Add policy lifecycle graph coverage for templates, versions, approvals, attestations, reviews, exceptions, and reminders.
- Add a bounded /grc/policy-lifecycle aggregate read with policy work queue, summaries, reminders, templates, and control/evidence mappings.
- Register the aggregate in the GRC report catalog and document the event/API contract.

## Validation
- go test ./internal/sourceprojection ./internal/bootstrap ./internal/grccatalog -count=1
- make sourcegen-check
- make policy-rule-check
- make detection-catalog-check
- make control-index-check
- make docs-drift-check
- make changed-check

---

<a href="https://app.devin.ai/review/writer/cerebro/pull/1499" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: internal/findings/policy_rule_catalog_ai_gen.go
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/findings/policy_rule_catalog_ai_gen.go:1]: Validated this path with `make contracts-check` and `make changed-check`. The generated policy rule catalogs, public detection catalog, control index, OpenAPI, docs drift, MCP contracts, and generated protobuf/SDK outputs are all clean against the YAML/source contracts; no hand-only generated drift remains.

File: internal/bootstrap/grc_policy_lifecycle.go [hunk 1]
```diff
@@ -0,0 +1,1294 @@
...
+	}
+	entityRows, err := store.ExecuteReadCypher(ctx, ports.CypherQueryRequest{
+		Query: grcPolicyLifecycleEntitiesQuery,
+		Params: map[string]any{
+			"tenant_id":    scope.TenantID,
+			"entity_types": grcPolicyLifecycleEntityTypes,
+			"limit":        limit,
+		},
+		RowLimit: limit,
+	})
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/bootstrap/grc_policy_lifecycle.go]: Fixed in the latest head: `grcpolicylifecycle.Scope` now carries `SourceID` and `RuntimeID`, both entity and relation Cypher reads apply those filters, and the endpoint test asserts the params and query predicates are present.

File: internal/bootstrap/grc_policy_lifecycle.go [hunk 2]
```diff
@@ -0,0 +1,1294 @@
...
+}
+
+func grcPolicyPendingStatus(status string) bool {
+	normalized := strings.ToLower(strings.TrimSpace(status))
+	return normalized == "" || normalized == "pending" || normalized == "requested" || normalized == "in_review" || normalized == "due" || normalized == "overdue"
+}
+
+func grcPolicyExceptionOpen(status string) bool {
+	return !grcPolicyClosedStatus(status)
+}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/bootstrap/grc_policy_lifecycle.go]: Fixed in the latest head: empty approval/review/attestation status no longer counts as pending work, empty exception status no longer counts as open work, and `internal/grcpolicylifecycle/lifecycle_test.go` covers both missing-status cases.

File: internal/grccatalog/catalog.go [hunk 3]
```diff
@@ -267,6 +267,19 @@ func buildCatalog() []Source {
...
+			Title:          "Policy lifecycle",
+			Description:    "Policy templates, versions, approvals, attestations, exceptions, reminders, and control mappings.",
+			Method:         "GET",
+			Path:           "/grc/policy-lifecycle",
+			Params:         nil,
+			Visualizations: []string{"table", "metric"},
+			DefaultLimit:   DefaultLimit,
+			MaxLimit:       MaxLimit,
+			CacheScope:     "graph",
+		},
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grccatalog/catalog.go]: Fixed in b5d6d157: the policy-lifecycle catalog source now allows the scoped parameters the handler actually consumes, `source_id` and `runtime_id`, while continuing to reject unsupported `runtime_ids`. Added catalog validation coverage for both accepted and rejected cases.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 4]
```diff
@@ -0,0 +1,1306 @@
...
+  AND e.entity_type IN $entity_types
+RETURN e.urn AS urn,
+       e.tenant_id AS tenant_id,
+       e.source_id AS source_id,
+       coalesce(e.runtime_id, '') AS runtime_id,
+       e.entity_type AS entity_type,
+       coalesce(e.label, e.urn) AS label,
+       coalesce(e.attributes_json, '{}') AS attributes_json
+ORDER BY e.entity_type, e.label, e.urn
+LIMIT $limit`
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go]: Fixed in b5d6d157: lifecycle entity reads now apply the request limit per lifecycle entity type through a per-type subquery, with row-limit math capped by the graph store ceiling so less-common objects are not starved by large policy sets. Added lifecycle tests for the query shape and row ceiling.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 5]
```diff
@@ -0,0 +1,1306 @@
...
+		EscalatedTo: grcPolicyActionActors(node.URN, relations, "escalated"),
+		DueAt:       grcPolicyAttr(node, "due_at", "scheduled_at"),
+		SentAt:      grcPolicyAttr(node, "sent_at"),
+	}
+}
+
+func grcPolicyFinalize(policy *grcPolicyLifecyclePolicy, now time.Time) {
+	sort.Slice(policy.Versions, func(i, j int) bool {
+		return grcPolicySortDate(policy.Versions[i].ApprovedAt, policy.Versions[i].CreatedAt, policy.Versions[i].EffectiveAt).After(grcPolicySortDate(policy.Versions[j].ApprovedAt, policy.Versions[j].CreatedAt, policy.Versions[j].EffectiveAt))
+	})
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go:684]: Fixed in 3e684835: version sorting now distinguishes missing dates from real dates, sorts dated versions first, and uses a stable ID/URN tiebreaker. Added a lifecycle test proving a dateless draft cannot override a dated approved version as latest.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 6]
```diff
@@ -0,0 +1,1306 @@
...
+		summary.Total++
+		if grcPolicyAcceptedStatus(item.Status) || item.AcceptedAt != "" {
+			summary.Accepted++
+			continue
+		}
+		if grcPolicyOverdue(item.DueAt, item.Status, now) {
+			summary.Overdue++
+		} else {
+			summary.Pending++
+		}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go]: Fixed in 3e684835: acceptance rollups now exclude closed unaccepted attestations from active totals and pending counts, while still counting accepted, overdue, and pending active attestations. Added coverage for accepted, pending, overdue, rejected, and expired states.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 7]
```diff
@@ -0,0 +1,1356 @@
...
+			policy.Reviews = append(policy.Reviews, item)
+			if policy.Reviewer == "" {
+				policy.Reviewer = firstNonEmpty(item.Reviewers...)
+			}
+			if policy.ReviewCadence == "" {
+				policy.ReviewCadence = item.Cadence
+			}
+			if policy.NextReviewDueAt == "" {
+				policy.NextReviewDueAt = item.ReviewDueAt
+			}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go]: Fixed in ed02275d: policy-level review metadata is now derived in `grcPolicyFinalize` after reviews are sorted, preserving explicit policy fields when present and otherwise selecting deterministic sorted review data. Added coverage for multiple reviews with different due dates.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 8]
```diff
@@ -0,0 +1,1356 @@
...
+	return items
+}
+
+func grcPolicyAcceptanceRollup(items []grcPolicyAcceptanceItem, now time.Time) grcPolicyAcceptanceSummary {
+	var summary grcPolicyAcceptanceSummary
+	for _, item := range items {
+		if grcPolicyAcceptedStatus(item.Status) || item.AcceptedAt != "" {
+			summary.Total++
+			summary.Accepted++
+			continue
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go]: Fixed in ed02275d: closed attestation statuses now take precedence over `accepted_at`, so rejected or expired records with stale accepted dates are excluded from active totals and accepted counts. Extended the rollup test to cover that edge case.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 9]
```diff
@@ -0,0 +1,1361 @@
...
+			if grcPolicyPast(item.ExpiresAt, now) {
+				summary.Expired++
+			}
+			continue
+		}
+		if grcPolicyPast(item.ExpiresAt, now) {
+			summary.Expired++
+			continue
+		}
+		summary.Active++
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go:889]: Fixed in 868b52d6: exception rollups now skip missing-status exceptions, matching the work queue `grcPolicyExceptionOpen` gate. Added a regression test for missing-status and active exception counts.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 10]
```diff
@@ -0,0 +1,1361 @@
...
+		return ""
+	}
+	if value := grcPolicyAttr(node, keys...); value != "" {
+		return value
+	}
+	if node.URN == "" {
+		return ""
+	}
+	parts := strings.Split(node.URN, ":")
+	return strings.TrimSpace(parts[len(parts)-1])
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go:1066]: Fixed in 868b52d6: policy nodes without `policy_id` now fall back to the full policy URN for their policy ID, preventing same-last-segment collisions. Added coverage for two distinct policy URNs that share the same final segment.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 11]
```diff
@@ -0,0 +1,1361 @@
...
+		}
+		if grcPolicyAcceptedStatus(item.Status) || item.AcceptedAt != "" {
+			summary.Total++
+			summary.Accepted++
+			continue
+		}
+		if grcPolicyOverdue(item.DueAt, item.Status, now) {
+			summary.Total++
+			summary.Overdue++
+		} else if grcPolicyPendingStatus(item.Status) {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go]: Fixed in 868b52d6: non-empty, non-accepted, non-closed attestation statuses now count as active pending attestations and appear in the work queue, while missing status remains excluded. Added rollup and work-queue regression tests for an unknown open status.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 12]
```diff
@@ -0,0 +1,1361 @@
...
+			To:       to,
+			Relation: grcPolicyRowString(row, "relation"),
+			Attrs:    grcPolicyAttrs(grcPolicyRowString(row, "relation_attributes_json")),
+		})
+	}
+
+	policyBuilders := map[string]*grcPolicyLifecyclePolicy{}
+	policyURNToID := map[string]string{}
+	for _, node := range nodes {
+		if !grcPolicyIsPolicyNode(node) {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go:424]: Fixed in 868b52d6: top-level policy promotion is now limited to policy nodes returned by the scoped entity query. Relation-only policy nodes can enrich relationships but are not promoted into the policy list. Added a regression test for relation enrichment with a foreign-source policy node.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 13]
```diff
@@ -0,0 +1,1388 @@
...
+func grcPolicyAcceptanceOpenStatus(status string) bool {
+	return strings.TrimSpace(status) != "" && !grcPolicyAcceptedStatus(status) && !grcPolicyClosedStatus(status)
+}
+
+func grcPolicyExceptionOpen(status string) bool {
+	return strings.TrimSpace(status) != "" && !grcPolicyClosedStatus(status)
+}
+
+func grcPolicyOverdue(rawDueAt string, status string, now time.Time) bool {
+	if grcPolicyAcceptedStatus(status) || grcPolicyClosedStatus(status) {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go]: Already fixed in 5d75f457: `grcPolicyOverdue` now returns false for empty status, so missing-status reviews do not inflate overdue counts or create work queue items.

File: internal/grcpolicylifecycle/lifecycle.go [hunk 14]
```diff
@@ -0,0 +1,1388 @@
...
+	return items
+}
+
+func grcPolicyAcceptanceRollup(items []grcPolicyAcceptanceItem, now time.Time) grcPolicyAcceptanceSummary {
+	var summary grcPolicyAcceptanceSummary
+	for _, item := range items {
+		if grcPolicyClosedStatus(item.Status) {
+			continue
+		}
+		if grcPolicyAcceptedStatus(item.Status) || item.AcceptedAt != "" {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/grcpolicylifecycle/lifecycle.go]: Already fixed in 5d75f457: `accepted_at` is only accepted as evidence when status is empty; explicit open or closed non-accepted statuses stay out of the accepted count. Focused lifecycle tests pass on the current head.
